src/tree/updater_quantile_hist.cc

```diff
@@ -98,18 +98,74 @@ bool QuantileHistMaker::UpdatePredictionCache(
 void QuantileHistMaker::Builder::SyncHistograms(
     int starting_index,
     int sync_count,
+    const GHistIndexMatrix &gmat,
     RegTree *p_tree) {
   builder_monitor_.Start("SyncHistograms");
-  this->histred_.Allreduce(hist_[starting_index].data(), hist_builder_.GetNumBins() * sync_count);
-  // use Subtraction Trick
-  for (auto const& node_pair : nodes_for_subtraction_trick_) {
-    hist_.AddHistRow(node_pair.first);
-    SubtractionTrick(hist_[node_pair.first], hist_[node_pair.second],
-                     hist_[(*p_tree)[node_pair.first].Parent()]);
+
+  const bool isDistributed = rabit::IsDistributed();
+  builder_monitor_.Start("AddHistRow");
+  for (auto const& node : nodes_for_subtraction_trick_) {
+    hist_.AddHistRow(node.nid);
+  }
+  builder_monitor_.Stop("AddHistRow");
+
+  const size_t nbins = gmat.cut.Ptrs().back();
+  common::BlockedSpace2d space(nodes_for_explicit_hist_build_.size(), [&](size_t node) {
+    return nbins;
+  }, 1024);
+
+  common::ParallelFor2d(space, [&](size_t node, common::Range1d r) {
+    const auto entry = nodes_for_explicit_hist_build_[node];
+    auto this_hist = hist_[entry.nid];
+    // Merging histograms from each thread into once
+    common::InitilizeHistByZeroes({this_hist.data() + r.begin(), this_hist.data() + r.end()});
+    hist_buffer_.ReduceHist(this_hist, node, r.begin(), r.end());
+
+    if (!(*p_tree)[entry.nid].IsRoot() && entry.sibling_nid > -1 && !isDistributed) {
+      auto parent_hist = hist_[(*p_tree)[entry.nid].Parent()];
+      auto sibling_hist = hist_[entry.sibling_nid];
+
+      for (bst_omp_uint bin_id = r.begin(); bin_id < r.end(); bin_id++) {
+        sibling_hist[bin_id].SetSubstract(parent_hist[bin_id], this_hist[bin_id]);
+      }
+    }
+  });
+
+  if (isDistributed) {
+    this->histred_.Allreduce(hist_[starting_index].data(), hist_builder_.GetNumBins() * sync_count);
+    // use Subtraction Trick
+    for (auto const& node : nodes_for_subtraction_trick_) {
+      SubtractionTrick(hist_[node.nid], hist_[node.sibling_nid],
+                       hist_[(*p_tree)[node.nid].Parent()]);
+    }
   }
+
   builder_monitor_.Stop("SyncHistograms");
 }
 
+void QuantileHistMaker::Builder::BuildHistogramsLossGuide(
+                        ExpandEntry entry,
+                        const GHistIndexMatrix &gmat,
+                        const GHistIndexBlockMatrix &gmatb,
+                        RegTree *p_tree,
+                        const std::vector<GradientPair> &gpair_h) {
+  nodes_for_explicit_hist_build_.clear();
+  nodes_for_subtraction_trick_.clear();
+  nodes_for_explicit_hist_build_.push_back(entry);
+
+  int starting_index = std::numeric_limits<int>::max();
+  int sync_count = 1;
```

Comment:
```suggestion
  int sync_count = 0;
```
The initial value should be 0 here. The following call to `BuildLocalHistograms()` will increment `sync_count` by 1, and we want to pass `sync_count=1` to `SyncHistograms()`.